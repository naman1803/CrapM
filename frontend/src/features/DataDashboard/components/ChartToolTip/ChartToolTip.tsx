import { Box, Grid2, Paper, Typography } from '@mui/material';
import { TooltipProps } from 'recharts';
import React from 'react';

export interface ChartToolTipProps extends TooltipProps<number, string> {
  labels: Record<string, string>;
}

/**
 * custom tooltip for recharts, insert into content prop field in recharts tooltip
 * labels = Record of field -> label on final tooltip
 */
const ChartToolTip: React.FC<ChartToolTipProps> = ({ payload, labels }) => {

  if (!payload || !payload.length) return null;

  return (
    <Box
      sx={{
        border: 1,
        borderColor: '#3a3a3a',
        borderRadius: 1,
      }}
      component={Paper}
      padding={'0.5rem'}
    >
      <Grid2>
        {Object.entries(payload[0].payload)
          // filter for only keys explicitly stated in `labels`
          .filter(([key]) => key in labels)
          .map(([key, value]) => (
            <Typography variant='h6' key={key}>
              {`${labels[key] || key}: ${value}`}
            </Typography>
          ))}
      </Grid2>
    </Box>
  );
};

export default ChartToolTip;
