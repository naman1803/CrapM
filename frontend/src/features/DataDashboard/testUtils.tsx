jest.mock('recharts', () => ({
  __esModule: true,
  ...jest.requireActual('recharts'),
  ResponsiveContainer: jest.fn().mockImplementation(({ children }) => (
    <div>
      <p>ResponsiveContainer</p>
      {children}
    </div>
  )),
}));
