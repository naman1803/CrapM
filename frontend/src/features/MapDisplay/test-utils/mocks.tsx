import {
  LakeGeoPoint,
  MapElement,
  MapElementType,
  SewageGeoPoint,
} from '../types/MapElements';

export const mockLakeElements: MapElement[] = [
  {
    id: 1,
    dataType: 'lake',
    isSelectable: true,
    elementType: MapElementType.Marker,
    dataPoint: {
      coordinates: [0, 0],
      id: 1,
      name: 'Grrrr Lake',
      additional_notes: 'This is a lake',
    } as LakeGeoPoint,
  },
  {
    id: 2,
    dataType: 'lake',
    isSelectable: true,
    elementType: MapElementType.Marker,
    dataPoint: {
      coordinates: [0, 0],
      id: 2,
      name: 'Obama Lake',
    } as LakeGeoPoint,
  },
];

export const mockSewageElements: MapElement[] = [
  {
    id: 1,
    dataType: 'sewage',
    isSelectable: true,
    elementType: MapElementType.Circle,
    dataPoint: {
      coordinates: [0, 0],
      id: 1,
      facility_owner: 'Eric',
      radius: 100,
      colour: 'Blue',
    } as SewageGeoPoint,
  },
  {
    id: 2,
    dataType: 'sewage',
    isSelectable: true,
    elementType: MapElementType.Circle,
    dataPoint: {
      coordinates: [0, 0],
      id: 2,
      facility_owner: 'Myron',
      radius: 100,
      colour: 'Red',
    } as SewageGeoPoint,
  },
];
