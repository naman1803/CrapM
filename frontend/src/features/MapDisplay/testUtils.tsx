import { LeafletEventHandlerFnMap, LatLngExpression } from 'leaflet';

jest.mock('react-leaflet', () => ({
  __esModule: true,
  ...jest.requireActual('leaflet'),
  useMap: () => ({
    flyTo: jest.fn(),
    getZoom: jest.fn().mockReturnValue(13), // Add getZoom method returning default zoom level
    on: jest.fn(), // Mock on method for event handling
    off: jest.fn(), // Mock off method for removing event listeners
  }),
  MapContainer: jest.fn().mockImplementation(({ children }) => (
    <div>
      <p>Map Container</p>
      {children}
    </div>
  )),
  Marker: jest
    .fn()
    .mockImplementation(
      ({
        position,
        children,
        key,
        eventHandlers,
      }: {
        eventHandlers: LeafletEventHandlerFnMap;
        position: LatLngExpression;
        key: any;
        children: React.ReactNode;
      }) => (
        <div key={key}>
          <p>{`Marker-${position.toString()}`}</p>
          {Object.entries(eventHandlers).map(([keyE, value]) => (
            <button key={keyE} onClick={value}>
              {`button-${keyE}-${position.toString()}`}
            </button>
          ))}
          {children}
        </div>
      ),
    ),
  Popup: jest.fn().mockImplementation(({ children }) => (
    <div>
      <p>Popup</p>
      {children}
    </div>
  )),
  TileLayer: jest.fn().mockImplementation(({ children }) => (
    <div>
      <p>TileLayer</p>
      {children}
    </div>
  )),
}));
