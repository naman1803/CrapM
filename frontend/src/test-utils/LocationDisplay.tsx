import { useLocation } from 'react-router-dom';
/**test component which displays the current url location.
 * use testid = location-display to get value of current location
 */
const LocationDisplay: React.FC = () => {
  const location = useLocation();

  return <div data-testid='location-display'>{location.pathname}</div>;
};
export default LocationDisplay;
