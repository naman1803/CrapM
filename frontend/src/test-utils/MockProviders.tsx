import { ThemeProvider } from '@mui/material';
import { MemoryRouter } from 'react-router-dom';
import { themeOptions } from '@app/theme';
import React from 'react';

// Wraps all providers for the app for testing
export const TestProviders = ({
  children,
  initialRoute = '/',
}: {
  children?: React.ReactNode;
  initialRoute?: string;
}) => {
  return (
    <ThemeProvider theme={themeOptions}>
      <MemoryRouter initialEntries={[initialRoute]}>{children}</MemoryRouter>
    </ThemeProvider>
  );
};
