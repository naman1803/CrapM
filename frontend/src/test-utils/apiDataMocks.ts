import { LakeData } from '@/types/LakeData';
import {
  SewageCoordData,
  SewageData,
  SewageSeverityData,
} from '@/types/SewageData';

export const mockSewageSeverityElement: SewageSeverityData = {
  id: 1,
  facilityOwner: 'Eric',
  contaminants: {
    brainrot: {
      averageSeverityScore: 1.0,
      overallSeverityLevel: 'Moderate',
      averagePercentExceedance: 100.0,
      records: [
        {
          severityScore: 1.0,
          severityLevel: 'Moderate',
          date: '2021-01-01',
        },
        {
          severityScore: 2.0,
          severityLevel: 'High',
          date: '2021-01-02',
        },
      ],
    },
    takingDumb: {
      averageSeverityScore: 2.0,
      overallSeverityLevel: 'High',
      averagePercentExceedance: 200.0,
      records: [
        {
          severityScore: 0.5,
          severityLevel: 'High',
          date: '2021-01-01',
        },
        {
          severityScore: 2.0,
          severityLevel: 'High',
          date: '2021-01-02',
        },
      ],
    },
  },
};

export const mockLakeApiData: LakeData[] = [
  {
    id: 1,
    lakeName: 'Lake Ontario',
    latitude: 45.2,
    longitude: -78.94,
    additionalNotes: null,
  },
  {
    id: 2,
    lakeName: 'Lake Erie',
    latitude: 46.47,
    longitude: -82.78,
    additionalNotes: null,
  },
];

export const mockSewageApiDataSent = [
  {
    id: 1,
    facilityOwner: 'Mock Facility 1',
    latitude: 44.49,
    longitude: -77.69,
    contaminant: 'BIOCHEMICAL OXYGEN DEMAND',
    radius: 22.539892,
    colour: "Green"
  },
  {
    id: 2,
    facilityOwner: 'Mock Facility 2',
    latitude: 44.68,
    longitude: -79.23,
    contaminant: 'PH - HIGH',
    radius: 22.524761,
    colour: "Green"
  },
];
export const mockSewageApiData: SewageCoordData[] = [
  {
    id: 1,
    facilityOwner: 'Mock Facility 1',
    latitude: 44.49,
    longitude: -77.69,
    radius: 22.539892,
    colour: "Green"
  },
  {
    id: 2,
    facilityOwner: 'Mock Facility 2',
    latitude: 44.68,
    longitude: -79.23,
    radius: 22.524761,
    colour: "Green"
  },
];

export const mockSewageSeverityApiData: SewageSeverityData[] = [
  {
    id: 1,
    facilityOwner: 'Mock Facility 1',
    contaminants: {
      'BIOCHEMICAL OXYGEN DEMAND': {
        averageSeverityScore: 0.77,
        overallSeverityLevel: 'Low',
        averagePercentExceedance: 177.0,
        records: [
          {
            severityScore: 0.77,
            severityLevel: 'Low',
            date: '2016-08-18',
          },
        ],
      },
    },
  },
  {
    id: 2,
    facilityOwner: 'Mock Facility 2',
    contaminants: {
      'PH - HIGH': {
        averageSeverityScore: 0.95,
        overallSeverityLevel: 'Low',
        averagePercentExceedance: 195.4,
        records: [
          {
            severityScore: 0.95,
            severityLevel: 'Low',
            date: '2016-08-30',
          },
        ],
      },
      'BIOCHEMICAL OXYGEN DEMAND': {
        averageSeverityScore: 0.58,
        overallSeverityLevel: 'Low',
        averagePercentExceedance: 157.8,
        records: [
          {
            severityScore: 0.58,
            severityLevel: 'Low',
            date: '2016-08-23',
          },
        ],
      },
    },
  },
];

export const mockSewageAllApiData: SewageData[] = [
  {
    id: 1,
    facilityOwner: 'Mock Facility 1',
    siteAddress: '123 Mock St',
    contaminant: 'BIOCHEMICAL OXYGEN DEMAND',
    typeOfExceedance: 'Legislation Non-Compliance',
    exceedanceStartDate: '2016-08-18T00:00:00',
    exceedanceEndDate: '2016-08-18T00:00:00',
    contaminantLimit: 50.0,
    contaminantUnit: 'mg/L',
    limitFrequency: 'any',
    noOfExceedances: 1,
    quantityMaximum: 88.5,
    quantityMinimum: null,
  },
  {
    id: 2,
    facilityOwner: 'Mock Facility 2',
    siteAddress: '1234 Mock St',
    contaminant: 'BIOCHEMICAL OXYGEN DEMAND',
    typeOfExceedance: 'Legislation Non-Compliance',
    exceedanceStartDate: '2016-08-23T00:00:00',
    exceedanceEndDate: '2016-08-23T00:00:00',
    contaminantLimit: 50.0,
    contaminantUnit: 'mg/L',
    limitFrequency: 'any',
    noOfExceedances: 1,
    quantityMaximum: 78.9,
    quantityMinimum: null,
  },
  {
    id: 3,
    facilityOwner: 'Mock Facility 2',
    siteAddress: '1234 Mock St',
    contaminant: 'PH - HIGH',
    typeOfExceedance: 'Legislation Non-Compliance',
    exceedanceStartDate: '2016-08-30T00:00:00',
    exceedanceEndDate: '2016-08-30T00:00:00',
    contaminantLimit: 50.0,
    contaminantUnit: 'mg/L',
    limitFrequency: 'any',
    noOfExceedances: 1,
    quantityMaximum: 97.7,
    quantityMinimum: null,
  },
];
