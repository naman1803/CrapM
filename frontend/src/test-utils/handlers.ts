import { http, HttpResponse } from 'msw';
import {
  mockLakeApiData,
  mockSewageAllApiData,
  mockSewageApiDataSent,
  mockSewageSeverityApiData,
} from './apiDataMocks';

const handlers = [
  http.get('/api/maps/lakes/all', () => {
    return HttpResponse.json(mockLakeApiData);
  }),
  http.get('/api/maps/sewage/all', () => {
    return HttpResponse.json(mockSewageApiDataSent);
  }),
  http.get('/api/data/sewage/all', () => {
    return HttpResponse.json(mockSewageAllApiData);
  }),
  ...mockSewageSeverityApiData.map((severityData) =>
    http.get(`/api/data/sewage/severity/${severityData.id}`, () => {
      return HttpResponse.json(severityData);
    }),
  ),
];
export default handlers;
