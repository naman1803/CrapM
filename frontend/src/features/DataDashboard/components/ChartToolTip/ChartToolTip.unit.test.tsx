import './../../testUtils';

import { TestProviders } from "@/test-utils/MockProviders";
import { render } from "@testing-library/react";
import ChartToolTip, { ChartToolTipProps } from './ChartToolTip';

const mockChartToolTipData: ChartToolTipProps = {
    payload: [
        {
            payload: {
                SeverityScore: 1,
                SillyScore: 2,
                FunScore: 3,
            },
        },
    ],
    labels: {
        SeverityScore: 'Severity Score',
        SillyScore: 'Silly Score',
        FunScore: 'Fun Score',
    },
};

describe('ChartToolTip', () => {
    it('should render the tool tip', () => {
        const { asFragment } = renderChartToolTip(mockChartToolTipData);
        expect(asFragment()).toMatchSnapshot();
    });
});

const renderChartToolTip = (props: ChartToolTipProps) => render(
    <TestProviders>
        <ChartToolTip {...props} />
    </TestProviders>
);
