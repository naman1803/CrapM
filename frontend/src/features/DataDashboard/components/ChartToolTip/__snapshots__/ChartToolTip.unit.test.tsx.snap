// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartToolTip should render the tool tip 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiBox-root css-1u5qmo5-MuiPaper-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12); --Paper-overlay: linear-gradient(rgba(255, 255, 255, 0.051), rgba(255, 255, 255, 0.051));"
  >
    <div
      class="MuiGrid2-root MuiGrid2-direction-xs-row css-1fzlhpv-MuiGrid2-root"
    >
      <h6
        class="MuiTypography-root MuiTypography-h6 css-rc2ge0-MuiTypography-root"
      >
        Severity Score: 1
      </h6>
      <h6
        class="MuiTypography-root MuiTypography-h6 css-rc2ge0-MuiTypography-root"
      >
        Silly Score: 2
      </h6>
      <h6
        class="MuiTypography-root MuiTypography-h6 css-rc2ge0-MuiTypography-root"
      >
        Fun Score: 3
      </h6>
    </div>
  </div>
</DocumentFragment>
`;
